adapter/k8s.go

```diff
@@ -0,0 +1,372 @@
+		if err != nil {
+			err = gherrors.Wrap(err, "unable to retrieve the resource with a matching name, while attempting to apply the config")
+			logrus.Error(err)
+			return nil, err
+		}
+	}
+	logrus.Infof("Retrieved Resource of type: %s and name: %s", data.GetKind(), data.GetName())
+	return data1, nil
+}
+
+func (h *BaseHandler) updateResource(ctx context.Context, res schema.GroupVersionResource, data *unstructured.Unstructured) error {
+	if _, err := h.DynamicKubeClient.Resource(res).Namespace(data.GetNamespace()).Update(ctx, data, metav1.UpdateOptions{}); err != nil {
+		err = gherrors.Wrap(err, "unable to update resource with the given name, attempting operation without namespace")
+		logrus.Warn(err)
+
+		if _, err = h.DynamicKubeClient.Resource(res).Update(ctx, data, metav1.UpdateOptions{}); err != nil {
+			err = gherrors.Wrap(err, "unable to update resource with the given name, while attempting to apply the config")
+			logrus.Error(err)
+			return err
+		}
+	}
+	logrus.Infof("Updated Resource of type: %s and name: %s", data.GetKind(), data.GetName())
+	return nil
+}
+
+func (h *BaseHandler) applyConfigChange(ctx context.Context, yamlFileContents, namespace string, isDelete, isCustomOp bool) error {
+	yamls, err := h.splitYAML(yamlFileContents)
+	if err != nil {
+		err = gherrors.Wrap(err, "error while splitting yaml")
+		logrus.Error(err)
+		return err
+	}
+	for _, yml := range yamls {
+		if strings.TrimSpace(yml) != "" {
+			if err := h.applyRulePayload(ctx, namespace, []byte(yml), isDelete, isCustomOp); err != nil {
+				errStr := strings.TrimSpace(err.Error())
+				if isDelete {
+					if strings.HasSuffix(errStr, "not found") ||
+						strings.HasSuffix(errStr, "the server could not find the requested resource") {
+						continue
+					}
+				} else {
+					if strings.HasSuffix(errStr, "already exists") {
+						continue
+					}
+				}
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (h *BaseHandler) applyRulePayload(ctx context.Context, namespace string, newBytes []byte, isDelete, isCustomOp bool) error {
+	if h.DynamicKubeClient == nil {
+		return errors.New("mesh client has not been created")
+	}
+	jsonBytes, err := yaml.YAMLToJSON(newBytes)
+	if err != nil {
+		err = gherrors.Wrapf(err, "unable to convert yaml to json")
+		logrus.Error(err)
+		return err
+	}
+	if len(jsonBytes) > 5 { // attempting to skip 'null' json
+		data := &unstructured.Unstructured{}
+		err = data.UnmarshalJSON(jsonBytes)
+		if err != nil {
+			err = gherrors.Wrapf(err, "unable to unmarshal json created from yaml")
+			logrus.Error(err)
+			return err
+		}
+		if data.IsList() {
+			err = data.EachListItem(func(r runtime.Object) error {
+				dataL, _ := r.(*unstructured.Unstructured)
+				return h.executeRule(ctx, dataL, namespace, isDelete, isCustomOp)
+			})
+			return err
+		}
+		return h.executeRule(ctx, data, namespace, isDelete, isCustomOp)
+	}
+	return nil
+}
+
+func (h *BaseHandler) splitYAML(yamlContents string) ([]string, error) {
```

Comment:
This function may not work well on `linkerd-adapter` maybe we can optimize it.

Reply:
Out of curiosity, why might it not work well?
Also, I'm slightly surprised that there is no library that allows to split a multi line string on arbitrary delimiter - at least I couldn't find any. Or specifically for YAML.
We should probably borrow the tests from apimachinery as well, especially if we optimize this function.

Reply:
As we know the Linkerd2 with many of CRDs, while we split the `yaml`, there is some resource missed and made the deployments belong to Linkerd2 like `tap` or other may not run to expected status

Reply:
Yep, I want try to find something interesting in the `kubectl` repo, but it's not happened. If I remember right, I saw `yaml` tools in `apimachinery` but may not includes spilt.

Reply:
Thanks for the clarification about the Linkerd2 issue. 

Yes, it's from apimachinery (https://github.com/kubernetes/apimachinery/blob/master/pkg/util/yaml/decoder.go) Unfortunately, splitYAML isn't public.